### ISSUE TITLE
Problem with DTK build

### ISSUE THREAD

**Issue by keitat** [trilinos/Trilinos#3071]
ECP-xSDK team has been working on type 2 package build and additional packages (DTK, etc.) for the fall release.  We have observed a build error when DTK (by ORNL) is enabled.

(gcc 7.1.0)
This is the current build setting in Spack (xsdk/pacakges.py).
` depends_on('trilinos@develop+hypre+superlu-dist+metis+hdf5~mumps+boost~suite-sparse+tpetra+nox+ifpack2+zoltan2+amesos2+dtk~exodus',`

But get:

```
1 error found in build log:
135 -- Setting Trilinos_ENABLE_STKSearchUtil=OFF because STKSearchUtil has a required library dependence on disabled package STKMesh
136 -- Setting Trilinos_ENABLE_STKTools=OFF because STKTools has a required library dependence on disabled package STKMesh
137 -- Setting Trilinos_ENABLE_PanzerAdaptersSTK=OFF because PanzerAdaptersSTK has a required library dependence on disabled package STKMesh
138 -- Setting Trilinos_ENABLE_STKUnit_test_utils=OFF because STKUnit_test_utils has a required library dependence on disabled package STKIO
139 -- Setting STKSearch_ENABLE_TESTS=OFF because STKSearch has a required test/example dependence on disabled package STKUnit_test_utils
140 -- Setting STKSearch_ENABLE_EXAMPLES=OFF because STKSearch has a required test/example dependence on disabled package STKUnit_test_utils
>> 141 CMake Error at cmake/tribits/core/package_arch/TribitsAdjustPackageEnables.cmake:1098 (MESSAGE):
142 ***
143 *** ERROR: Setting Trilinos_ENABLE_DataTransferKit=OFF which was 'ON' because DataTransferKit has a required library dependence on disabled package D
ataTransferKitDiscretization!
144 ***
145 
146 Call Stack (most recent call first):
147 cmake/tribits/core/package_arch/TribitsAdjustPackageEnables.cmake:1181 (TRIBITS_PRIVATE_PRINT_DISABLE)
```

I think some packages are missing in order to integrate DTK.  Please let me know if you need more information.

**Comment by mhoemmen** [comment]
Incidentally, I would recommend GCC 7.2.0, as we will likely need to support that for the Sierra computer.

**Comment by mhoemmen** [comment]
So, what happens when you enable DataTransferKitDiscretization?

**Comment by mhoemmen** [comment]
@rppawlo Are you in charge of DTK?

**Comment by keitat** [comment]
Is "DataTransferKitDiscretization" enabled by Trilinos CMAKE or DTK?  I guess I am missing the point.

**Comment by mhoemmen** [comment]
@keitat Try setting `Trilinos_ENABLE_DataTransferKitDiscretization:BOOL=ON` perhaps?

**Comment by keitat** [comment]
OK. This will require some changes in Trilinos package.py in Spack.

**Comment by mhoemmen** [comment]
@keitat btw do you work on maintaining the Trilinos part of Spack?  IBM folks have reported issues with that.  See e.g., https://github.com/trilinos/Trilinos/issues/2938

**Comment by keitat** [comment]
Yes, I have to take care of this, too.  Mark, I need your help to tackle these issues.

**Comment by mhoemmen** [comment]
@keitat wrote:
> Mark, I need your help to tackle these issues.

Let's talk more over e-mail.  (Did you realize that we maintain three separate build systems for Trilinos: bjam (for Sierra), CMake, and Spack?)

**Comment by Rombur** [comment]
DTK now depends on `Intrepid2`. I guess that the spack formula is not up to date.

**Comment by mhoemmen** [comment]
Trilinos' CMake logic normally would deduce what packages need enabling.  I suspect the problem is that people turn off a bunch of packages explicitly.

**Comment by aprokop** [comment]
Spack does not play well together with Trilinos CMake logic. It's more of a hardcoded thing where you decide the exact packages to build a priori, and then cross the fingers that it builds. I added a bunch of checks to Spack Trilinos' `package.py` based on required dependencies in Trilinos' `Dependencies.cmake` for different packages. I forgot to do the same for the extras like `DTK`.

**Comment by aprokop** [comment]
I submitted spack/spack#8660. Now it will break early, during spack dependency resolution rather than during the Trilinos configuration.

@keitat You will still need to add `+intrepid2` to your trilinos spec.

**Comment by mhoemmen** [comment]
@aprokop wrote:
> Spack does not play well together with Trilinos CMake logic.

Hm, that sounds familiar, there may be another build system like this ;-)

> P.S. I kind of "maintain" the Trilinos package in spack. You can ping me when you have issues.

This is good to know; thanks! :-D

**Comment by rppawlo** [comment]
> @rppawlo Are you in charge of DTK?

@mhoemmen - no, @sslattery is DTK lead.
